### PR TITLE
feat: Worksheet 1 badges logic

### PR DESCRIPTION
**Pull request by faucomte97**

### CODE REVIEW

**Review by dionizh — CHANGES_REQUESTED** [review]
> Reviewed 14 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 10 unresolved discussions (waiting on @faucomte97)

___
*[`.github/workflows/ci.yml` line 29 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yGizIDm-La7PXkrpZ:-N1yGizJ5ysGNhkyEw4s:b-vmx4u5) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/.github/workflows/ci.yml#L29)):*
> ```YAML
>           pipenv install --dev --system
>           pip install -U "git+https://github.com/ocadotechnology/codeforlife-portal@aimmo_badges_db#egg=cfl-common&subdirectory=cfl_common"
>           pip install -U "git+https://github.com/ocadotechnology/codeforlife-portal@aimmo_badges_db#egg=codeforlife-portal"
> ```

This is to remove later?
___
*[`aimmo/views.py` line 63 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yGtgJBfTFVW3muFn1:-N1yGtgJBfTFVW3muFn2:b2ohuyk) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L63)):*
> ```Python
> def badges(request, id):
>     if not request.user:
>         print("no user")
> ```

Is this a debug print? If not it'd be more useful to have more info?
___
*[`aimmo/views.py` line 67 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yH-YD4k1ZeKyWKQ0j:-N1yH-YD4k1ZeKyWKQ0k:br3zhre) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L67)):*
> ```Python
>     game = get_object_or_404(Game, id=id)
>     if not game.can_user_play(request.user):
>         print("user can't play")
> ```

Same question/comment as above.
___
*[`aimmo/views.py` line 73 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yHE7l-OQhlhSczbMQ:-N1yHE7l-OQhlhSczbMR:b4rkgvh) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L73)):*
> ```Python
>     except Avatar.DoesNotExist:
>         avatar = create_avatar_for_user(request.user, id)
>     avatar_user_profile = UserProfile.objects.get(user=avatar.owner)
> ```

Can you put space after this line. It's all getting a bit cramped.
___
*[`aimmo/tests/test_views.py` line 550 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yHgeh1OwRWtb8IzH8:-N1yHgeh1OwRWtb8IzH9:baxc1ic) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/tests/test_views.py#L550)):*
> ```Python
>         game3 = Game(id=3, name="test", game_class=klass3, status=Game.RUNNING)
>         game3.save()
>         Game(id=4, name="test", game_class=klass4, status=Game.STOPPED)
> ```

Is this changing the game? Maybe put comment as it's not clear why you don't assign this to a variable. And the line after this is saving game3 instead of game4? Actually I remember Kamil was changing the line below to `game4.save()` at the worksheet4 branch. We're all a bit confused.
___
*[`game_frontend/src/pyodide/badges.ts` line 7 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yOwuS4DwVs3K1xnZi:-N1yOwuS4DwVs3K1xnZj:bvsv6v5) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/badges.ts#L7)):*
> ```XML
>   const userPythonCode = userCode.replace(/\s*#.*/gm, "") // Remove all comment lines from the user's code
>   const badgesPerWorksheet = [
>     {id: 1, worksheetID: 1, trigger: badge1Trigger(result)},
> ```

I wonder whether id is better as string, so you don't need to convert it to string again below. Also in the future, it'd be possible for us to add new tasks on a worksheet, so we could have 1, 1b, 2, 3, 3b for instance (this is assuming that worksheet 2 has badge 4, 5, 6, etc and by adding more task on one worksheet, we don't want to shift the badge numbers for all the worksheets after that) -- if that makes sense..
___
*[`game_frontend/src/pyodide/badges.ts` line 35 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yKAswAQCEgkG5AZWN:-N1yKAswAQCEgkG5AZWO:b-uf9j54) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/badges.ts#L35)):*
> ```XML
> 
> function badge3Trigger(result: any, userPythonCode: string, gameState: any, playerAvatarId: number): boolean {
>   // Check code contains keywords
> ```

Can you put more comment what badge 3 trigger is supposed to check? Below you can see `check next cell is available to move to` so it's not just checking keywords.
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 26 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLbQf-WfsB1aRVFih:-N1yLbQf-WfsB1aRVFii:b-goxjxt) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L26)):*
> ```XML
>   playerAvatarId: number
> ): Promise<string> {
>   return runBadgeCheck(() => worker.checkIfBadgeEarned(badges, result, userCode, gameState, playerAvatarId))
> ```

Does `checkIfBadgeEarned` check all three badges every turn? Can't really see where a check result is being stored.
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 70 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLB3A94Ed_Upksl__:-N1yLB3A94Ed_Upksl_a:b-2rnp10) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L70)):*
> ```XML
>     return func()
>   } else {
>     return Promise.resolve("error")
> ```

Should this error message be more descriptive?
___
*[`game_frontend/src/redux/features/AvatarWorker/epics.js` line 86 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yO8NN8414rv5oOXZl:-N1yO8NN8414rv5oOXZm:b-t8bpk8) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/redux/features/AvatarWorker/epics.js#L86)):*
> ```JavaScript
>   )
> 
> const getBadgesEpic = (action$, state$, { api }) =>
> ```

Can you put comments on these 3 functions? I'm not really sure what's going on but comments will help 😆

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 10 unresolved discussions (waiting on @dionizh)

___
*[`.github/workflows/ci.yml` line 29 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yGizIDm-La7PXkrpZ:-N2CZImL8WMXfUt1fxyO:b-lrxfy3) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/.github/workflows/ci.yml#L29)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

This is to remove later?
</blockquote></details>

Yes, after I merge in the portal PR
___
*[`aimmo/views.py` line 63 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yGtgJBfTFVW3muFn1:-N2CxQkAEu782SKxm7El:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L63)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Is this a debug print? If not it'd be more useful to have more info?
</blockquote></details>

Done.
___
*[`aimmo/views.py` line 67 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yH-YD4k1ZeKyWKQ0j:-N2CxQy5-UiOTguyYTVC:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L67)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Same question/comment as above.
</blockquote></details>

Done.
___
*[`aimmo/views.py` line 73 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yHE7l-OQhlhSczbMQ:-N2Ci-5s65BNbOY1SGVD:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/views.py#L73)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Can you put space after this line. It's all getting a bit cramped.
</blockquote></details>

Done.
___
*[`aimmo/tests/test_views.py` line 550 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yHgeh1OwRWtb8IzH8:-N2Crk0fBMnxwZlexGZ6:b-b5swbc) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/aimmo/tests/test_views.py#L550)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Is this changing the game? Maybe put comment as it's not clear why you don't assign this to a variable. And the line after this is saving game3 instead of game4? Actually I remember Kamil was changing the line below to `game4.save()` at the worksheet4 branch. We're all a bit confused.
</blockquote></details>

Oh yeah, didn't even notice that. It's weird yeah... I think we can even just call `save` after the bracket and not have to use a variable anyway...
___
*[`game_frontend/src/pyodide/badges.ts` line 7 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yOwuS4DwVs3K1xnZi:-N2CsMTW37gtdliJkNSq:b-393rwu) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/badges.ts#L7)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

I wonder whether id is better as string, so you don't need to convert it to string again below. Also in the future, it'd be possible for us to add new tasks on a worksheet, so we could have 1, 1b, 2, 3, 3b for instance (this is assuming that worksheet 2 has badge 4, 5, 6, etc and by adding more task on one worksheet, we don't want to shift the badge numbers for all the worksheets after that) -- if that makes sense..
</blockquote></details>

Does it actually make a difference? I would still need to use the ``` symbol to concatenate the `:` and the `,`.  Otherwise, it could be `badges += badge.worksheetID + ":" + badge.id + ","` which doesn't seem as nice, right?
___
*[`game_frontend/src/pyodide/badges.ts` line 35 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yKAswAQCEgkG5AZWN:-N2CshWG1SqnU02EzBwM:b5mf7sk) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/badges.ts#L35)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Can you put more comment what badge 3 trigger is supposed to check? Below you can see `check next cell is available to move to` so it's not just checking keywords.
</blockquote></details>

Ah yes, sorry, this wasn't meant to be a generic comment for the whole check, just the first step.
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 26 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLbQf-WfsB1aRVFih:-N2Ct4Z_DhCmpfqXulXa:b-k5neei) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L26)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Does `checkIfBadgeEarned` check all three badges every turn? Can't really see where a check result is being stored.
</blockquote></details>

It checks the badges for the active worksheet, and checks if the badge has already been earned on not (it's in the if statement in the loop inside the function). And it's not every turn, only when the user clicks the Run Code button.

The reason you can't see how the result is being used is because of redux and how horrible it is. Basically the Promise returns the `result`, and `checkIfBadgeEarned` is called in the Epic where it is used. That epic is triggered only when the user presses the Run Code button (when the Avatar's code is updated).
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 70 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLB3A94Ed_Upksl__:-N2Cu4-_DrCogakz7vEI:b-aokmix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L70)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Should this error message be more descriptive?
</blockquote></details>

I need to rethink this actually, it might be a problem and might erase the badge data. I'll come back to you for that one.
___
*[`game_frontend/src/redux/features/AvatarWorker/epics.js` line 86 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yO8NN8414rv5oOXZl:-N2Cx9f03RH74UjWO6aN:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/redux/features/AvatarWorker/epics.js#L86)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Can you put comments on these 3 functions? I'm not really sure what's going on but comments will help 😆
</blockquote></details>

Done.

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: 7 of 14 files reviewed, 10 unresolved discussions (waiting on @dionizh)

___
*[`game_frontend/src/pyodide/badges.ts` line 7 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yOwuS4DwVs3K1xnZi:-N2GkNv61zYhUfExUg8J:b8u26nd) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/badges.ts#L7)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Does it actually make a difference? I would still need to use the ``` symbol to concatenate the `:` and the `,`.  Otherwise, it could be `badges += badge.worksheetID + ":" + badge.id + ","` which doesn't seem as nice, right?
</blockquote></details>

I just realised what you meant, Done :slightly_smiling_face: 
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 70 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLB3A94Ed_Upksl__:-N2GkTccB8sVGYYtQWwP:b-o6jsr6) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L70)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

I need to rethink this actually, it might be a problem and might erase the badge data. I'll come back to you for that one.
</blockquote></details>

It was an issue - fixed it by rejecting the Promise instead, raising an error, and raising a Failure type in that case so it doesn't actually POST the error result to the DB.

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 2 unresolved discussions (waiting on @dionizh)

___
*[`aimmo/tests/test_views.py` line 577 at r5](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N2HEUoj3JXPmpOs8Q4R:-N2I-IIb00IaCcq_N1Uy:b-5k0jkz) ([raw file](https://github.com/ocadotechnology/aimmo/blob/9e68162ba339a116d879d27481c1cf4faa712f7f/aimmo/tests/test_views.py#L577)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Is the id supposed to be `1:1`?
</blockquote></details>

No, the `id` here represents the `id` of the Game :slightly_smiling_face: 
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 70 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLB3A94Ed_Upksl__:-N2I10I42xBOpNJW8Gbv:b-2ofh3u) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L70)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

Do we need to return an error if the worker is not ready? Can we not just pass/ignore until the worker is ready? (i.e. can we remove the else?)
</blockquote></details>

Turns out we didn't need that function altogether :man_shrugging:

**Review by dionizh — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 1 unresolved discussion (waiting on @faucomte97)

___
*[`aimmo/views.py` line 91 at r6](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N2Ly7GQ6-LtXMExWj0X:-N2Ly7GQ6-LtXMExWj0Y:b3eti1i) ([raw file](https://github.com/ocadotechnology/aimmo/blob/5896c5e60f606b495fb154eafb701e7855a1dcef/aimmo/views.py#L91)):*
> ```Python
>             aimmo_badges = ""
> 
>         return JsonResponse({"badges": aimmo_badges})
> ```

We can shorten this to `aimmo_badges or ""`
___
*[`game_frontend/src/pyodide/pyodideRunner.ts` line 70 at r2](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N1yLB3A94Ed_Upksl__:-N2LyLgl1XcRn_VYgeR4:b2nhjic) ([raw file](https://github.com/ocadotechnology/aimmo/blob/08b848d15d3612d252f80c7d7a540ca01f375958/game_frontend/src/pyodide/pyodideRunner.ts#L70)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Turns out we didn't need that function altogether :man_shrugging: 
</blockquote></details>

Lol

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: 12 of 14 files reviewed, 1 unresolved discussion (waiting on @dionizh)

___
*[`aimmo/views.py` line 91 at r6](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N2Ly7GQ6-LtXMExWj0X:-N2M0Nu11-ZNbAmlBHzA:b-896fix) ([raw file](https://github.com/ocadotechnology/aimmo/blob/5896c5e60f606b495fb154eafb701e7855a1dcef/aimmo/views.py#L91)):*
<details><summary><i>Previously, dionizh (Dioni Zhong) wrote…</i></summary><blockquote>

We can shorten this to `aimmo_badges or ""`
</blockquote></details>

Done.

**Review by dionizh — COMMENTED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 1 unresolved discussion (waiting on @dionizh)

**Review by dionizh — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-:-N2MFw1ICCvqBNT0CCYg:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @dionizh)

**Review by dionizh — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @faucomte97)

**Review by dionizh — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: all files reviewed, 1 unresolved discussion (waiting on @faucomte97)

___
*[`aimmo/views.py` line 85 at r9](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-N2MhjlF1-0EjLcWBy9O:-N2MhjlF1-0EjLcWBy9P:b-xn5m4) ([raw file](https://github.com/ocadotechnology/aimmo/blob/38980165f292d87fc970bd4e03137f8b64297f69/aimmo/views.py#L85)):*
> ```Python
>         pattern = re.compile("^([1-9]:\d+,)*$")
> 
>         if re.match(pattern, earned_badges):
> ```

This can be shorten to `re.match("^([1-9]:\d+,)*$", earned_bages)`



_Code quote:_
```Python
^([1-9]:\d+,)*$
```

**Review by dionizh — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655#-:-N2MjaE55A2TpXzsnTAy:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @dionizh)

**Review by dionizh — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/aimmo/1655)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @faucomte97)
